### PR TITLE
8299179: ArrayFill with store on backedge needs to reduce length by 1

### PR DESCRIPTION
**Pull request by eme64**
**Context**
`ArrayFill` replaces `CountedLoop` like this:
`for(int i = 5; i < 15; i++) { arr[i] = 88; }`
with
`jint_fill(arr, 5, len, 88)` (CallLeafNoFPNode, intrinsified)
where `len = limit - init`.
This is implemented for `byte`, `short` and `int` (`long` not yet supported).

**Solution**
This assumes that the `store` is in the body of the loop. In rare cases, the `store` is pinned to the backedge, and so it is executed after the exit check, hence in all iterations except the last one. In those cases we need to set `len = limit - init - 1`.

**Details**
_How does a store land on the backedge?_
`test_002` was first reported to me. I reduced it down to a conceptually easier `test_102`.
We have a loop, where the first block only has a single successor. This block contains the `store` that it eventually detected as a `ArrayFill`. This block passes the checks of `clone_loop_head` during parsing: the block is copied before the loop and onto the backedge.

**FYI**
I saw that we seem to currently mostly unroll such loops, instead of applying the `ArrayFill`. Follow up RFE [JDK-8299808](https://bugs.openjdk.org/browse/JDK-8299808).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299179](https://bugs.openjdk.org/browse/JDK-8299179): ArrayFill with store on backedge needs to reduce length by 1

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [6cd85641](https://git.openjdk.org/jdk/pull/11904/files/6cd85641527f0589a5e4ba7dbaa2fa4bffde494f)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11904/head:pull/11904` \
`$ git checkout pull/11904`

Update a local copy of the PR: \
`$ git checkout pull/11904` \
`$ git pull https://git.openjdk.org/jdk pull/11904/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11904`

View PR using the GUI difftool: \
`$ git pr show -t 11904`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11904.diff">https://git.openjdk.org/jdk/pull/11904.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -4483,5 +4496,12 @@ bool PhaseIdealLoop::intrinsify_fill(IdealLoopTree* lpt) {
     _igvn.replace_node(n, C->top());
   }
 
+#ifndef PRODUCT
+    if (TraceOptimizeFill) {
+      tty->print("ArrayFill call   ");
+      call->dump();
+    }
+#endif
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopTransform.cpp:4504]:
```suggestion
#ifndef PRODUCT
  if (TraceOptimizeFill) {
    tty->print("ArrayFill call   ");
    call->dump();
  }
#endif
```

**Review by TobiHartmann — APPROVED** [review]
Nice analysis and test. Looks good to me.

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 2]
```diff
@@ -4381,6 +4381,19 @@ bool PhaseIdealLoop::intrinsify_fill(IdealLoopTree* lpt) {
   Node* len = new SubINode(head->limit(), head->init_trip());
   _igvn.register_new_node_with_optimizer(len);
 
+  // If the store is on the backedge, it is not executed in the last
+  // iteration, and we must subtract 1 from the len.
+  Node* backedge = head->loopexit()->proj_out(1);
+  if (store->in(0) == backedge) {
+    len = new SubINode(len, _igvn.intcon(1));
+    _igvn.register_new_node_with_optimizer(len);
```
Comment (dean-long) [inline src/hotspot/share/opto/loopTransform.cpp:4389]: How about moving the earlier call to register_new_node_with_optimizer to after this block, so it is only called once?

Comment (eme64) [reply src/hotspot/share/opto/loopTransform.cpp:4389]: Hi @dean-long . Just discussed it with @TobiHartmann . We have two new nodes here, one to compute
`len = limit - init`,
the other to decrement the len by 1:
`len = len - 1`.
Both are new nodes, and need to be registered with igvn. So we think it is correct the way it is now.

Comment (dean-long) [reply src/hotspot/share/opto/loopTransform.cpp:4389]: OK, that makes sense.

**Comment by eme64** [comment]
Thanks @TobiHartmann @vnkozlov @dean-long for the help and reviews!
/integrate

**Pull request closed without merging**